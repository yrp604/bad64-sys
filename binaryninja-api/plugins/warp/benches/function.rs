use binaryninja::binary_view::BinaryViewExt;
use binaryninja::headless::Session;
use criterion::{criterion_group, criterion_main, Criterion};
use rayon::prelude::*;
use std::path::PathBuf;
use warp_ninja::build_function;

// These are the target files present in OUT_DIR
// Add the files to fixtures/bin
static TARGET_FILES: [&str; 1] = ["atox.obj"];

pub fn function_benchmark(c: &mut Criterion) {
    let session = Session::new().expect("Failed to initialize session");
    let out_dir = env!("OUT_DIR").parse::<PathBuf>().unwrap();
    for file_name in TARGET_FILES {
        let path = out_dir.join(file_name);
        let view = session.load(&path).expect("Failed to load view");

        let functions: Vec<_> = view.functions().iter().map(|f| f.to_owned()).collect();
        // Bench all functions sequentially
        c.bench_function("signature all functions", |b| {
            b.iter(|| {
                for func in &functions {
                    let _ = build_function(&func, || func.lifted_il().ok(), false);
                }
            })
        });

        // Bench all functions in parallel
        c.bench_function("signature all functions rayon", |b| {
            b.iter(|| {
                functions
                    .par_iter()
                    .filter_map(|func| {
                        Some(build_function(
                            func.as_ref(),
                            || func.lifted_il().ok(),
                            false,
                        ))
                    })
                    .collect::<Vec<_>>()
            })
        });
    }
}

criterion_group!(benches, function_benchmark);
criterion_main!(benches);
