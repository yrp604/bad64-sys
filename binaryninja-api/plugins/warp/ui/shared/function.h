#pragma once
#include <QStandardItemModel>
#include <QTableView>

#include "binaryninjaapi.h"
#include "constraint.h"
#include "filter.h"
#include "misc.h"
#include "warp.h"

class WarpFunctionItem : public QStandardItem
{
    Warp::Ref<Warp::Function> m_function;

    // Optional attached data used to show/manage the function.
    Warp::Ref<Warp::Container> m_container;
    std::optional<Warp::Source> m_source;

public:
    WarpFunctionItem(Warp::Ref<Warp::Function> function,
                     BinaryNinja::Ref<BinaryNinja::Function> analysisFunction);

    void SetContainer(const Warp::Ref<Warp::Container> &container);

    void SetSource(Warp::Source source);

    Warp::Ref<Warp::Function> GetFunction() { return m_function; }
    Warp::Ref<Warp::Container> GetContainer() { return m_container; }
    std::optional<Warp::Source> GetSource() { return m_source; }
};

class WarpFunctionItemModel : public QStandardItemModel
{
    Q_OBJECT

    // The current matched function, used to highlight currently.
    Warp::Ref<Warp::Function> m_matchedFunction;

    // Mapping of function start address to the row index.
    // This is used to identify unique functions for updating instead of resetting the entire model.
    std::unordered_map<uint64_t, int> m_insertableFunctionRows;

public:
    WarpFunctionItemModel(const QStringList &labels, QObject *parent);

    static constexpr int COL_FUNCTION_ITEM = 0;
    static constexpr int COL_ADDRESS_ITEM = 1;

    void AppendFunction(WarpFunctionItem *item);

    void InsertFunction(uint64_t address, WarpFunctionItem *item);

    WarpFunctionItem *GetItem(const QModelIndex &index) const;

    std::optional<uint64_t> GetAddress(const QModelIndex &index) const;

    QVariant data(const QModelIndex &index, int role) const override;

    void SetMatchedFunction(const Warp::Ref<Warp::Function> &matchedFunction)
    {
        Warp::Ref<Warp::Function> previousMatchedFunction = m_matchedFunction;
        m_matchedFunction = matchedFunction;

        // Make sure to refresh the highlights so we don't keep the highlights from the previous function.
        if (previousMatchedFunction)
        {
            const QModelIndex topLeft = index(0, 0);
            const QModelIndex bottomRight = index(rowCount() - 1, 0);
            emit dataChanged(topLeft, bottomRight);
        }
    }
};

class WarpFunctionFilterModel : public QSortFilterProxyModel
{
    Q_OBJECT

public:
    WarpFunctionFilterModel(QObject *parent): QSortFilterProxyModel(parent)
    {
    }

    ~WarpFunctionFilterModel() override = default;

    bool filterAcceptsRow(int sourceRow, const QModelIndex &sourceParent) const override;

    bool lessThan(const QModelIndex &sourceLeft, const QModelIndex &sourceRight) const override;
};

class WarpFunctionTableWidget : public QWidget, public FilterTarget
{
    Q_OBJECT

    QTableView *m_table;
    WarpFunctionItemModel *m_model;
    WarpFunctionFilterModel *m_proxyModel;
    FilterEdit *m_filterEdit;
    FilteredView *m_filterView;
    QMenu *m_contextMenu;
    std::map<QString, std::function<void(WarpFunctionItem *, std::optional<uint64_t>)> > m_contextMenuActions;

public:
    explicit WarpFunctionTableWidget(QWidget *parent = nullptr);

    // TODO: Invert this and provide OnCallback functions that wrap the connect call.
    QTableView *GetTableView() const { return m_table; }
    WarpFunctionItemModel *GetModel() const { return m_model; }
    WarpFunctionFilterModel *GetProxyModel() const { return m_proxyModel; }

    void RegisterContextMenuAction(const QString &name,
                                   const std::function<void(WarpFunctionItem *, std::optional<uint64_t>)> &callback);

    void SetFunctions(QVector<WarpFunctionItem *> functions);

    void InsertFunction(uint64_t address, WarpFunctionItem *function);

    void setFilter(const std::string &) override;

    void scrollToFirstItem() override
    {
    }

    void scrollToCurrentItem() override
    {
    }

    void selectFirstItem() override
    {
    }

    void activateFirstItem() override
    {
    }
};

class WarpFunctionInfoWidget : public QWidget
{
    Q_OBJECT

    Warp::Ref<Warp::Function> m_function;
    BinaryNinja::Ref<BinaryNinja::Function> m_analysisFunction;

    // Optionally provide this information to show the source information.
    Warp::Ref<Warp::Container> m_container;
    std::string source;

    WarpConstraintTableWidget *m_constraintsTable;

    QTableView *m_commentsTable;
    QStandardItemModel *m_commentsModel;

    QTableView *m_variablesTable;

public:
    explicit WarpFunctionInfoWidget(QWidget *parent = nullptr);

    Warp::Ref<Warp::Function> GetFunction() { return m_function; }
    void SetFunction(Warp::Ref<Warp::Function> function) { m_function = function; };

    void SetAnalysisFunction(BinaryNinja::Ref<BinaryNinja::Function> analysisFunction)
    {
        m_analysisFunction = analysisFunction;
    };
    BinaryNinja::Ref<BinaryNinja::Function> GetAnalysisFunction() { return m_analysisFunction; }

    // TODO: Make this private?
    void UpdateInfo();
};
