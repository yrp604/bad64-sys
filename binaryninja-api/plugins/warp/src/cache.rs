pub mod container;
pub mod function;
pub mod guid;
pub mod type_reference;

pub use function::*;
pub use guid::*;
pub use type_reference::*;

use binaryninja::binary_view::{BinaryView, BinaryViewExt};
use binaryninja::function::Function as BNFunction;
use binaryninja::rc::Guard;
use binaryninja::rc::Ref as BNRef;
use binaryninja::ObjectDestructor;
use std::hash::{DefaultHasher, Hash, Hasher};

pub fn register_cache_destructor() {
    pub static mut CACHE_DESTRUCTOR: CacheDestructor = CacheDestructor;
    #[allow(static_mut_refs)]
    // SAFETY: This can be done as the backing data is an opaque ZST.
    unsafe {
        CACHE_DESTRUCTOR.register()
    };
}

/// A unique view ID, used for caching.
#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct ViewID(u64);

impl From<&BinaryView> for ViewID {
    fn from(value: &BinaryView) -> Self {
        let mut hasher = DefaultHasher::new();
        hasher.write_u64(value.original_image_base());
        hasher.write_usize(value.file().session_id());
        Self(hasher.finish())
    }
}

impl From<BNRef<BinaryView>> for ViewID {
    fn from(value: BNRef<BinaryView>) -> Self {
        Self::from(value.as_ref())
    }
}

impl From<Guard<'_, BinaryView>> for ViewID {
    fn from(value: Guard<'_, BinaryView>) -> Self {
        Self::from(value.as_ref())
    }
}

/// A unique function ID, used for caching.
#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct FunctionID(u64);

impl From<&BNFunction> for FunctionID {
    fn from(value: &BNFunction) -> Self {
        let mut hasher = DefaultHasher::new();
        hasher.write_u64(value.start());
        hasher.write_u64(value.lowest_address());
        hasher.write_u64(value.highest_address());
        Self(hasher.finish())
    }
}

impl From<BNRef<BNFunction>> for FunctionID {
    fn from(value: BNRef<BNFunction>) -> Self {
        Self::from(value.as_ref())
    }
}

impl From<Guard<'_, BNFunction>> for FunctionID {
    fn from(value: Guard<'_, BNFunction>) -> Self {
        Self::from(value.as_ref())
    }
}

pub struct CacheDestructor;

impl ObjectDestructor for CacheDestructor {
    fn destruct_view(&self, view: &BinaryView) {
        clear_type_ref_cache(view);
        log::debug!("Removed WARP caches for {:?}", view.file().filename());
    }
}
