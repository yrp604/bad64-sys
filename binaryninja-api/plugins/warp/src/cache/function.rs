use crate::convert::{comment_to_bn_comment, to_bn_symbol_at_address};
use binaryninja::binary_view::BinaryViewExt;
use binaryninja::function::{Function as BNFunction, FunctionUpdateType};
use binaryninja::symbol::SymbolType;
use warp::signature::function::Function;

// TODO: Rename this?
/// Inserts a function match into the cache. This also has the side effect of setting persisted function
/// information, such as the matched function symbol.
///
/// IMPORTANT: This will mark the function as needing updates, if you intend to fill in functions with
/// no match (i.e. `None`), then you must change this function to prevent marking that as needing updates.
/// However, it's perfectly valid to remove a match and need to update the function still, so be careful.
pub fn insert_cached_function_match(function: &BNFunction, matched_function: Option<Function>) {
    let view = function.view();
    let function_start = function.start();
    // NOTE: If we expect to run match_function multiple times on a function, we should move this elsewhere.
    // Mark the function as needing updates so that reanalysis occurs on the function, and we apply the match.
    function.mark_updates_required(FunctionUpdateType::FullAutoFunctionUpdate);
    if let Some(auto_sym) = view.symbol_by_address(function_start) {
        // TODO: If we ever create non library function symbols we will need to remove this check (see: `to_bn_symbol_at_address`).
        if auto_sym.sym_type() == SymbolType::LibraryFunction {
            // NOTE: This will also mark for full auto function update, one thing to note is that the
            // requirement to call this is that this function not be called in the associated function's analysis.
            view.undefine_auto_symbol(&auto_sym);
        }
    }
    match matched_function {
        Some(matched_function) => {
            // Define the new matched function symbol, this can safely be done here as the symbol itself
            // will be persisted, unlike function type information or variable information.
            let new_sym = to_bn_symbol_at_address(&view, &matched_function.symbol, function_start);
            view.define_auto_symbol(&new_sym);
            // TODO: How to clear the comments? They are just persisted.
            // TODO: Also they generate an undo action, i hate implicit undo actions so much.
            for comment in &matched_function.comments {
                let bn_comment = comment_to_bn_comment(&function, comment.clone());
                function.set_comment_at(bn_comment.addr, &bn_comment.comment);
            }
            function.store_metadata("warp_matched_function", &matched_function.to_bytes(), false);
        }
        None => {
            function.remove_metadata("warp_matched_function");
        }
    }
}

// TODO: This does allocations, and for every reanalysis.
pub fn try_cached_function_match(function: &BNFunction) -> Option<Function> {
    let metadata = function.query_metadata("warp_matched_function")?;
    let raw_metadata = metadata.get_raw()?;
    Function::from_bytes(&raw_metadata)
}
