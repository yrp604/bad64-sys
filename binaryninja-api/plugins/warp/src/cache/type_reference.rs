use crate::cache::ViewID;
use crate::convert::from_bn_type_internal;
use binaryninja::binary_view::BinaryView;
use binaryninja::confidence::MAX_CONFIDENCE;
use binaryninja::rc::Guard;
use binaryninja::rc::Ref as BNRef;
use binaryninja::types::NamedTypeReference as BNNamedTypeReference;
use dashmap::mapref::one::Ref;
use dashmap::DashMap;
use std::collections::HashSet;
use std::hash::{DefaultHasher, Hasher};
use std::sync::OnceLock;
use warp::r#type::ComputedType;

pub static TYPE_REF_CACHE: OnceLock<DashMap<ViewID, TypeRefCache>> = OnceLock::new();

pub fn clear_type_ref_cache(view: &BinaryView) {
    let view_id = ViewID::from(view);
    if let Some(cache) = TYPE_REF_CACHE.get() {
        cache.remove(&view_id);
    }
}

pub fn cached_type_reference(
    view: &BinaryView,
    visited_refs: &mut HashSet<TypeRefID>,
    type_ref: &BNNamedTypeReference,
) -> Option<ComputedType> {
    let view_id = ViewID::from(view);
    let type_ref_cache = TYPE_REF_CACHE.get_or_init(Default::default);
    match type_ref_cache.get(&view_id) {
        Some(cache) => cache.cached_type_reference(view, visited_refs, type_ref),
        None => {
            let cache = TypeRefCache::default();
            let ntr = cache.cached_type_reference(view, visited_refs, type_ref);
            type_ref_cache.insert(view_id, cache);
            ntr
        }
    }
}

pub fn cached_type_references(view: &BinaryView) -> Option<Ref<ViewID, TypeRefCache>> {
    let view_id = ViewID::from(view);
    let type_ref_cache = TYPE_REF_CACHE.get_or_init(Default::default);
    type_ref_cache.get(&view_id)
}

#[derive(Clone, Debug, Default)]
pub struct TypeRefCache {
    pub cache: DashMap<TypeRefID, Option<ComputedType>>,
}

impl TypeRefCache {
    /// NOTE: No self-referential type must be used on this function.
    pub fn cached_type_reference(
        &self,
        view: &BinaryView,
        visited_refs: &mut HashSet<TypeRefID>,
        type_ref: &BNNamedTypeReference,
    ) -> Option<ComputedType> {
        let ntr_id = TypeRefID::from(type_ref);
        match self.cache.get(&ntr_id) {
            Some(cache) => cache.to_owned(),
            None => match type_ref.target(view) {
                Some(raw_ty) => {
                    let computed_ty = ComputedType::new(from_bn_type_internal(
                        view,
                        visited_refs,
                        &raw_ty,
                        MAX_CONFIDENCE,
                    ));
                    self.cache
                        .entry(ntr_id)
                        .insert(Some(computed_ty))
                        .to_owned()
                }
                None => self.cache.entry(ntr_id).insert(None).to_owned(),
            },
        }
    }
}

/// A unique named type reference ID, used for caching.
#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct TypeRefID(u64);

impl From<&BNNamedTypeReference> for TypeRefID {
    fn from(value: &BNNamedTypeReference) -> Self {
        let mut hasher = DefaultHasher::new();
        hasher.write(value.id().as_bytes());
        Self(hasher.finish())
    }
}

impl From<BNRef<BNNamedTypeReference>> for TypeRefID {
    fn from(value: BNRef<BNNamedTypeReference>) -> Self {
        Self::from(value.as_ref())
    }
}

impl From<Guard<'_, BNNamedTypeReference>> for TypeRefID {
    fn from(value: Guard<'_, BNNamedTypeReference>) -> Self {
        Self::from(value.as_ref())
    }
}
