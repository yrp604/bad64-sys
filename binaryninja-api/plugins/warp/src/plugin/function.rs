use crate::cache::{cached_function_guid, try_cached_function_guid};
use crate::{get_warp_include_tag_type, INCLUDE_TAG_NAME};
use binaryninja::background_task::BackgroundTask;
use binaryninja::binary_view::{BinaryView, BinaryViewExt};
use binaryninja::command::{Command, FunctionCommand};
use binaryninja::function::Function;
use binaryninja::rc::Guard;
use rayon::iter::ParallelIterator;
use std::thread;
use warp::signature::function::FunctionGUID;

pub struct IncludeFunction;

impl FunctionCommand for IncludeFunction {
    fn action(&self, view: &BinaryView, func: &Function) {
        let sym_name = func.symbol().short_name();
        let sym_name_str = sym_name.to_string_lossy();
        let should_add_tag = func.function_tags(None, Some(INCLUDE_TAG_NAME)).is_empty();
        let insert_tag_type = get_warp_include_tag_type(view);
        match should_add_tag {
            true => {
                log::info!(
                    "Including selected function '{}' at 0x{:x}",
                    sym_name_str,
                    func.start()
                );
                func.add_tag(&insert_tag_type, "", None, false, None);
            }
            false => {
                log::info!(
                    "Removing included function '{}' at 0x{:x}",
                    sym_name_str,
                    func.start()
                );
                func.remove_tags_of_type(&insert_tag_type, None, false, None);
            }
        }
    }

    fn valid(&self, _view: &BinaryView, _func: &Function) -> bool {
        // TODO: Only allow if the function is named?
        true
    }
}

pub struct CopyFunctionGUID;

impl FunctionCommand for CopyFunctionGUID {
    fn action(&self, _view: &BinaryView, func: &Function) {
        let Some(guid) = cached_function_guid(func, || func.lifted_il().ok()) else {
            log::error!("Could not get guid for copied function");
            return;
        };
        log::info!(
            "Function GUID for {:?}... {}",
            func.symbol().short_name(),
            guid
        );
        if let Ok(mut clipboard) = arboard::Clipboard::new() {
            let _ = clipboard.set_text(guid.to_string());
        }
    }

    fn valid(&self, _view: &BinaryView, _func: &Function) -> bool {
        true
    }
}

pub struct FindFunctionFromGUID;

impl Command for FindFunctionFromGUID {
    fn action(&self, view: &BinaryView) {
        let Some(guid_str) = binaryninja::interaction::get_text_line_input(
            "Function GUID",
            "Find Function from GUID",
        ) else {
            return;
        };

        let Ok(searched_guid) = guid_str.parse::<FunctionGUID>() else {
            log::error!("Failed to parse function guid... {}", guid_str);
            return;
        };

        log::info!("Searching functions for GUID... {}", searched_guid);
        let funcs = view.functions();
        let view = view.to_owned();
        thread::spawn(move || {
            let background_task = BackgroundTask::new(
                &format!("Searching functions for GUID... {}", searched_guid),
                false,
            );

            // Only run this for functions which have already generated a GUID.
            let matched: Vec<Guard<Function>> = funcs
                .par_iter()
                .filter(|func| {
                    try_cached_function_guid(func).is_some_and(|guid| guid == searched_guid)
                })
                .collect();

            if matched.is_empty() {
                log::info!("No matches found for GUID... {}", searched_guid);
            } else {
                for func in &matched {
                    // Also navigate the user, as that is probably what they want.
                    if matched.len() == 1 {
                        let current_view = view.file().current_view();
                        if view
                            .file()
                            .navigate_to(&current_view, func.start())
                            .is_err()
                        {
                            log::error!(
                                "Failed to navigate to found function 0x{:0x} in view {}",
                                func.start(),
                                current_view
                            );
                        }
                    }
                    log::info!("Match found at function... 0x{:0x}", func.start());
                }
            }

            background_task.finish();
        });
    }

    fn valid(&self, _view: &BinaryView) -> bool {
        true
    }
}
