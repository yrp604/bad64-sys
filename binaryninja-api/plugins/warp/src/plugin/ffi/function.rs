use crate::build_function;
use crate::cache::{insert_cached_function_match, try_cached_function_match};
use crate::convert::{to_bn_symbol_at_address, to_bn_type};
use crate::plugin::ffi::{BNWARPConstraint, BNWARPFunction, BNWARPFunctionGUID};
use binaryninja::function::Function;
use binaryninja::rc::Ref;
use binaryninja::string::BnString;
use binaryninjacore_sys::{BNFunction, BNSymbol, BNType};
use std::ffi::c_char;
use std::mem::ManuallyDrop;
use std::sync::Arc;
use warp::signature::comment::FunctionComment;

#[repr(C)]
pub struct BNWarpFunctionComment {
    pub text: *mut c_char,
    pub offset: i64,
}

impl BNWarpFunctionComment {
    /// Leaks the text string to be freed with BNWARPFreeFunctionComment
    pub fn from_owned(value: &FunctionComment) -> Self {
        let text = BnString::into_raw(BnString::new(&value.text));
        Self {
            text,
            offset: value.offset,
        }
    }

    pub unsafe fn free_raw(value: &Self) {
        unsafe { BnString::free_raw(value.text) }
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPGetFunction(
    analysis_function: *mut BNFunction,
) -> *mut BNWARPFunction {
    let function = Function::from_raw(analysis_function);
    match build_function(&function, || function.lifted_il().ok(), false) {
        Some(function) => Arc::into_raw(Arc::new(function)) as *mut BNWARPFunction,
        None => std::ptr::null_mut(),
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPGetMatchedFunction(
    analysis_function: *mut BNFunction,
) -> *mut BNWARPFunction {
    let function = Function::from_raw(analysis_function);
    match try_cached_function_match(&function) {
        Some(matched_function) => {
            let arc_matched_function = Arc::new(matched_function);
            // NOTE: Freed by BNWARPFreeFunctionReference
            Arc::into_raw(arc_matched_function) as *mut BNWARPFunction
        }
        None => std::ptr::null_mut(),
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionApply(
    function: *mut BNWARPFunction,
    analysis_function: *mut BNFunction,
) {
    let analysis_function = Function::from_raw(analysis_function);
    match function.is_null() {
        false => {
            // Set the matched function to `function`.
            let matched_function = ManuallyDrop::new(Arc::from_raw(function));
            insert_cached_function_match(
                &analysis_function,
                Some(matched_function.as_ref().clone()),
            )
        }
        true => {
            // We are removing the previous match.
            insert_cached_function_match(&analysis_function, None)
        }
    };
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetGUID(
    function: *mut BNWARPFunction,
) -> BNWARPFunctionGUID {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    function.guid
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetSymbol(
    function: *mut BNWARPFunction,
    analysis_function: *mut BNFunction,
) -> *mut BNSymbol {
    let analysis_function = Function::from_raw(analysis_function);
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let view = analysis_function.view();
    let address = analysis_function.symbol().address();
    let function_symbol = to_bn_symbol_at_address(&view, &function.symbol, address);
    // NOTE: The symbol ref has been pre-incremented for the caller.
    Ref::into_raw(function_symbol).handle
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetSymbolName(function: *mut BNWARPFunction) -> *mut c_char {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let bn_name = BnString::new(&function.symbol.name);
    // NOTE: The symbol name string to be freed by BNFreeString
    BnString::into_raw(bn_name)
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetType(
    function: *mut BNWARPFunction,
    analysis_function: *mut BNFunction,
) -> *mut BNType {
    let analysis_function = Function::from_raw(analysis_function);
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    match &function.ty {
        Some(func_ty) => {
            let arch = analysis_function.arch();
            let function_type = to_bn_type(Some(arch), func_ty);
            // NOTE: The type ref has been pre-incremented for the caller.
            unsafe { Ref::into_raw(function_type) }.handle
        }
        None => std::ptr::null_mut(),
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetConstraints(
    function: *mut BNWARPFunction,
    count: *mut usize,
) -> *mut BNWARPConstraint {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let raw_constraints: Box<[BNWARPConstraint]> = function
        .constraints
        .clone()
        .into_iter()
        .map(Into::into)
        .collect();
    *count = raw_constraints.len();
    let raw_constraints_ptr = Box::into_raw(raw_constraints);
    raw_constraints_ptr as *mut BNWARPConstraint
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetComments(
    function: *mut BNWARPFunction,
    count: *mut usize,
) -> *mut BNWarpFunctionComment {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let raw_comments: Box<[_]> = function
        .comments
        .iter()
        .map(BNWarpFunctionComment::from_owned)
        .collect();
    *count = raw_comments.len();
    let raw_comments_ptr = Box::into_raw(raw_comments);
    raw_comments_ptr as *mut BNWarpFunctionComment
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionsEqual(
    function_a: *mut BNWARPFunction,
    function_b: *mut BNWARPFunction,
) -> bool {
    // We do not own function so we should not drop.
    let function_a = ManuallyDrop::new(Arc::from_raw(function_a));
    // We do not own function so we should not drop.
    let function_b = ManuallyDrop::new(Arc::from_raw(function_b));
    function_a.eq(&function_b)
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPNewFunctionReference(
    function: *mut BNWARPFunction,
) -> *mut BNWARPFunction {
    Arc::increment_strong_count(function);
    function
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeFunctionReference(function: *mut BNWARPFunction) {
    if function.is_null() {
        return;
    }
    Arc::decrement_strong_count(function);
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeFunctionList(functions: *mut *mut BNWARPFunction, count: usize) {
    let functions_ptr = std::ptr::slice_from_raw_parts_mut(functions, count);
    let functions = Box::from_raw(functions_ptr);
    for function in functions {
        // NOTE: The functions themselves should also be arc.
        BNWARPFreeFunctionReference(function);
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeFunctionCommentList(
    comments: *mut BNWarpFunctionComment,
    count: usize,
) {
    let comments_ptr = std::ptr::slice_from_raw_parts_mut(comments, count);
    let comments = Box::from_raw(comments_ptr);
    for comment in &comments {
        BNWarpFunctionComment::free_raw(comment)
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeConstraintList(
    constraints: *mut BNWARPConstraint,
    count: usize,
) {
    let constraints_ptr = std::ptr::slice_from_raw_parts_mut(constraints, count);
    let _constraints = unsafe { Box::from_raw(constraints_ptr) };
}
