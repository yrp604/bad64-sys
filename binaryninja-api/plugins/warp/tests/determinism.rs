//! This tests to make sure the function GUIDs are stable.
use binaryninja::binary_view::BinaryViewExt;
use binaryninja::headless::Session;
use std::collections::BTreeMap;
use std::path::PathBuf;
use warp::signature::function::FunctionGUID;
use warp_ninja::cache::cached_function_guid;

// These are the target files present in OUT_DIR
// Add the files to fixtures/bin
static TARGET_FILES: [&str; 10] = [
    "_ctype.obj",
    "_fptostr.obj",
    "_mbslen.obj",
    "_memicmp.obj",
    "_strnicm.obj",
    "_wctype.obj",
    "atof.obj",
    "atoldbl.obj",
    "atox.obj",
    "ls",
];

#[test]
fn insta_signatures() {
    let session = Session::new().expect("Failed to initialize session");
    let out_dir = env!("OUT_DIR").parse::<PathBuf>().unwrap();
    for file_name in TARGET_FILES {
        let path = out_dir.join(file_name);
        let view = session.load(&path).expect("Failed to load view");
        let functions: BTreeMap<u64, FunctionGUID> = view
            .functions()
            .iter()
            .map(|f| {
                let guid = cached_function_guid(&f, || f.lifted_il().ok());
                (f.start(), guid.unwrap())
            })
            .collect();
        insta::assert_debug_snapshot!(file_name, functions);
    }
}
