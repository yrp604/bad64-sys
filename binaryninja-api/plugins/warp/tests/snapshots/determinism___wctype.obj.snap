---
source: plugins/warp/tests/determinism.rs
expression: functions
---
{
    181536: FunctionGUID {
        guid: 9350843d-64a5-59a9-b39d-490b424fc116,
    },
    181856: FunctionGUID {
        guid: 8a0c3535-d4db-5c0e-93ec-13830ae546d2,
    },
    182544: FunctionGUID {
        guid: a8036a14-9afb-59fe-ba38-8102a56c0e0d,
    },
    182896: FunctionGUID {
        guid: 119c08fa-5fb3-529e-9bc8-feb53495aeef,
    },
    183200: FunctionGUID {
        guid: 730daa1c-47e1-5bf8-a17c-03151d3fa093,
    },
    183536: FunctionGUID {
        guid: 24bb0d12-ee2d-5f3b-b7d6-cee173baf054,
    },
    183872: FunctionGUID {
        guid: 40f70a3b-90a4-57e5-9f47-58a555ea3431,
    },
    184208: FunctionGUID {
        guid: 01c1dec6-2bbe-5123-8ec5-1fe1cf7e1049,
    },
    184544: FunctionGUID {
        guid: 1aebf174-5039-53b1-916c-6a57b6f7897d,
    },
    184960: FunctionGUID {
        guid: 993a31a2-9b56-5dae-b3dd-e7296de852da,
    },
    185280: FunctionGUID {
        guid: d38af5e1-264d-57b8-b1ff-92b20257bd07,
    },
    185600: FunctionGUID {
        guid: 6d7fa497-54b7-5a83-a748-acd5edc2b8e4,
    },
    185952: FunctionGUID {
        guid: 565e1a72-6374-5ef6-85d9-37df7a16e0a2,
    },
    186272: FunctionGUID {
        guid: 40f70a3b-90a4-57e5-9f47-58a555ea3431,
    },
    186624: FunctionGUID {
        guid: 01c1dec6-2bbe-5123-8ec5-1fe1cf7e1049,
    },
    186976: FunctionGUID {
        guid: 6f47b85a-9d9d-543b-9fb1-9a7d99b63e1e,
    },
    187296: FunctionGUID {
        guid: 5c121dba-fe3e-5032-adea-46f41a52ea76,
    },
    187616: FunctionGUID {
        guid: a56c4f9f-3442-5c92-8e48-6165553b6996,
    },
    187936: FunctionGUID {
        guid: f8412b03-939f-58f5-8c09-189c7c8872f4,
    },
    188256: FunctionGUID {
        guid: 23bb8d8f-3805-5baa-8dd5-185ec947ca75,
    },
    188576: FunctionGUID {
        guid: 382b7353-4a13-566c-88f1-bc6b4aac31ee,
    },
    188896: FunctionGUID {
        guid: ca31b84d-b85d-588d-b761-532fc618cef6,
    },
    189216: FunctionGUID {
        guid: cb58263f-b52b-52d8-8a7a-583319ee9191,
    },
    189552: FunctionGUID {
        guid: 0d616fbd-f28e-5f57-8f3e-86c08d0462ba,
    },
    189856: FunctionGUID {
        guid: 993a31a2-9b56-5dae-b3dd-e7296de852da,
    },
    190160: FunctionGUID {
        guid: d38af5e1-264d-57b8-b1ff-92b20257bd07,
    },
    190464: FunctionGUID {
        guid: 1c72f175-686e-5387-b12a-19c1e8dd3151,
    },
    190768: FunctionGUID {
        guid: 6d7fa497-54b7-5a83-a748-acd5edc2b8e4,
    },
    191088: FunctionGUID {
        guid: 565e1a72-6374-5ef6-85d9-37df7a16e0a2,
    },
    191392: FunctionGUID {
        guid: 6f47b85a-9d9d-543b-9fb1-9a7d99b63e1e,
    },
    191696: FunctionGUID {
        guid: 5c121dba-fe3e-5032-adea-46f41a52ea76,
    },
    192000: FunctionGUID {
        guid: a56c4f9f-3442-5c92-8e48-6165553b6996,
    },
    192304: FunctionGUID {
        guid: f8412b03-939f-58f5-8c09-189c7c8872f4,
    },
    192608: FunctionGUID {
        guid: 23bb8d8f-3805-5baa-8dd5-185ec947ca75,
    },
    192912: FunctionGUID {
        guid: 382b7353-4a13-566c-88f1-bc6b4aac31ee,
    },
    193216: FunctionGUID {
        guid: ca31b84d-b85d-588d-b761-532fc618cef6,
    },
    193520: FunctionGUID {
        guid: cb58263f-b52b-52d8-8a7a-583319ee9191,
    },
}
